pkg/controller/logstorage/users/user_controller.go

```diff
@@ -245,3 +254,43 @@ func (r *UserController) createLinseedLogin(ctx context.Context, tenantID string
 
 	return nil
 }
+
+func (r *UserController) cleanupStaleUsers(ctx context.Context, logger logr.Logger) error {
+	esClient, err := r.esClientFn(r.client, ctx, relasticsearch.ElasticEndpoint())
+	if err != nil {
+		return fmt.Errorf("failed to connect to Elasticsearch - failed to create the Elasticsearch client")
+	}
+
+	// Each Elasticsearch instance is only connected to a single management cluster, but if that ever changes we'll
+	// need to rework this logic so that we're only deleting stale users created by the current management cluster.
+	allEsUsers, err := esClient.GetUsers(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch users from Elasticsearch")
+	}
+
+	tenants := operatorv1.TenantList{}
+	err = r.client.List(ctx, &tenants)
+	if err != nil {
+		return fmt.Errorf("failed to fetch TenantList")
+	}
+
+	for _, user := range allEsUsers {
+		if strings.HasPrefix(user.Username, utils.ElasticsearchUserNameLinseed) {
+			active := false
+			for _, t := range tenants.Items {
+				if strings.Contains(user.Username, t.Spec.ID) {
```

Comment:
I think it would be nice to have a symmetrical function to `formatName` called `parseName` that can take a given user name and return to component parts. That way we can do an exact match on the tenant ID instead of checking for a substring.

I think the approach you have here is probably safe, but I have a slight bit of FUD about potentially accidentally matching other users because of the way string matching works.

For example, imagine these two hypothetical usernames following the pattern `<name>_<tenantID>`:

1. Name="user", TenantID="tenant-id" -> `user_tenant-id`
2. Name="user_tenant", TenantID="id" -> `user_tenant_id`

If we do a substring match on the tenant ID "id", it will match both of those user names. So the second one will never be cleaned up so long as the first one does. This is a bit of a contrived scenario, but I think it's still worth considering since we don't know what the full set of potential tenant IDs in the future might be.

Reply:
This is the formatName I'm talking about: https://github.com/tigera/operator/blob/b4c9fc66998fc750dd59a3a79ef2ff788024529b/pkg/controller/utils/elasticsearch.go#L152-L168

Reply:
Some immediate consequences that fall out of introducing a `parseName` function are that we now must restrict the character set of `name`, `tenantID`, and `clusterName` fields to not contain "-" or "_". Is that a restriction we're ok with?

Reply:
Hmm, yeah I think that's not a restriction we can enforce, unfortunately.

I wonder if that means we need to change our `formatName` implementation. Or if there's other metadata associated with the user / role in ES we can use to more precisely determine the tenant ID and cluster name (other than the username itself)?